CRE-293 Add Metering Detail to Capabilities Response

EasterTheBunny opened the pull request:
Every node must independently report metering information while the engine expects an aggregated list of all nodes. The client expects an encoded `CapabilityResponse` as the payload with exactly one metering report value. These values are then aggregated for the expected number of response messages and included as metadata in the response forwarded to the engine. This approach isolates the aggregation logic to a single location in the code.

<!--- Does this work have a corresponding ticket?  Please link it here as well as one of:
    - the PR title
    - branch name
    - commit message
[LINK-777](https://smartcontract-it.atlassian.net/browse/LINK-777)
--> 
[CRE-293](https://smartcontract-it.atlassian.net/browse/CRE-293)

### Requires
<!--- Does this work depend on other open PRs? Please list them.
- https://github.com/smartcontractkit/chainlink-common/pull/7777777
-->
- https://github.com/smartcontractkit/chainlink-common/pull/1080

Follow on to https://github.com/smartcontractkit/chainlink-common/pull/1099


[CRE-293]: https://smartcontract-it.atlassian.net/browse/CRE-293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cedric-cordenier commented on core/services/workflows/engine.go:
```diff
@@ -863,9 +863,27 @@ func (e *Engine) workerForStepRequest(ctx context.Context, msg stepRequest) {
+		}
+	}
+
+	e.meterReport.SetStep(MeteringReportStepRef(stepState.Ref), meteringSteps)
```
Hmmmm you don't need to add the execution ID anywhere?

EasterTheBunny replied:
I am going to wait for https://github.com/smartcontractkit/chainlink/pull/16923 to be merged before updating this. I think that should satisfy the execution ID requirement, unless you think it should go in the report itself.

cedric-cordenier commented on core/capabilities/remote/executable/request/client_request.go:
```diff
@@ -190,15 +190,39 @@ func (c *ClientRequest) OnMessage(_ context.Context, msg *types.MessageBody) err
 
 	if msg.Error == types.Error_OK {
-		responseID := sha256.Sum256(msg.Payload)
+		responseID, metadata, err := c.getMessageHashAndMetadata(msg)
```
Can you add a comment here explaining why we are having to handle metering differently/separately here?

cedric-cordenier commented on core/services/workflows/engine.go:
```diff
@@ -1044,10 +1061,10 @@ func (e *Engine) executeStep(ctx context.Context, lggr logger.Logger, msg stepRe
 	}
 
-	return inputsMap, output.Value, err
+	return inputsMap, output.Value, output.Metadata, err
```
Might as well just return the response in the second field :)

EasterTheBunny replied:
Well I was thinking that too but then I thought keeping values the same `*values.Map, values.Value, ...`. I overthought it.

patrickhuie19 commented on core/capabilities/remote/executable/request/client_request.go:
```diff
@@ -93,50 +87,16 @@ func newClientRequest(ctx context.Context, lggr logger.Logger, requestID string,
 
 		wg.Add(1)
-		go func(innerCtx context.Context, peerID ragep2ptypes.PeerID, delay time.Duration) {
+		go func(ctx context.Context, peerID ragep2ptypes.PeerID, delay time.Duration) {
```
IFIUC this ctx no longer has the effectiveTimeout baked in. Why remove that?

EasterTheBunny replied:
This `ctx` is scoped to the gofunc and actually a little bit lower the select case should be using `<-ctx.Done()`. Since the `ctxWithCancel` is being passed to the gofunc, `ctx` should not shadow the parent function `ctx`.

patrickhuie19 replied:
Hmm - we want the effectiveTimeout encoded here somewhere as dispatcher.Send() is wired through to the base protocol layer and we don't want this gr to leak. 

There seems to be a few things missing here, so I agree with your sense that overal context wiring here needs to be cleaned up. It looks like `cancelFn` is never called, and that the wg for the go funcs is unused.

EasterTheBunny replied:
`cancelFn` gets called when `Cancel` is called. Should the `requestTimeout` be wrapped in a new context within `newClientRequest`? However, I think these changes are engine specific and might need to be part of another PR.

EasterTheBunny replied:
Wait, I see what you're talking about. How did I delete so much? This must have been a rebase mistake.

patrickhuie19 replied:
Hmm - I think I would
* remove the wg. It's unused and I don't see any upside to blocking here. (@cedric-cordenier correct me here?). Update the logging to`context done, not sending capability request to peer`
* Keep the effective timeout in the context we pass to these grs.

EasterTheBunny replied:
The rebase was botched. I corrected the issue and force pushed. @patrickhuie19 it might be best to take this as a conversation on another PR, just to not conflate the two subjects.

patrickhuie19 commented on core/services/workflows/engine.go:
```diff
@@ -854,27 +854,51 @@ func (e *Engine) workerForStepRequest(ctx context.Context, msg stepRequest) {
+			l.Error(fmt.Sprintf("failed to set metering report step for ref %s: %s", stepState.Ref, err))
+		}
+	} else {
+		l.Warn("no metering report found for " + msg.state.ExecutionID)
```
Could we also emit a metric? A gauge value would work here - 0 if metering report included, 1 if metering report not included

cedric-cordenier replied:
@patrickhuie19 Just curious, why a gauge as opposed to a count? The latter feels more intuitive since you'll get a record of the number of events

IMO a counter with two statuses (present/missing) would be the most natural thing to include here

patrickhuie19 replied:
The goal was to alert if there's a regression. At the time I thought we want every capability to meter, so a gauge felt sufficient to me.

I've updated to a counter. It'll give us more control if only some capabilities meter. Thanks :)

EasterTheBunny committed:
Add Metering Detail to Capabilities Response

Every node must independently report metering information while the engine expects an aggregated list of all nodes. The
client expects an encoded `CapabilityResponse` as the payload with exactly one metering report value. These values are
then aggregated for the expected number of response messages and included as metadata in the response forwarded to the
engine. This approach isolates the aggregation logic to a single location in the code.

EasterTheBunny committed:
simplify function output

EasterTheBunny committed:
code cleanup, linting, and go mod

EasterTheBunny committed:
address feedback and fix CI

patrickhuie19 committed:
Merge branch 'develop' into CRE-293/add-cost-field-capability-response

cedric-cordenier reviewed (changes requested):
No unit tests 😨

patrickhuie19 committed:
adding with_metering_metadata test

patrickhuie19 commented:
> 😨

9cc1ca0c17844169e5b759d48b0cd3c09f7747ff

cedric-cordenier commented on core/capabilities/remote/executable/request/client_request.go:
```diff
@@ -229,25 +228,50 @@ func (c *ClientRequest) OnMessage(_ context.Context, msg *types.MessageBody) err
+			return fmt.Errorf("failed to get message hash: %w", err)
+		}
 
-		lggr := c.lggr.With("responseID", hex.EncodeToString(responseID[:]), "count", c.responseIDCount[responseID], "requiredCount", c.requiredIdenticalResponses, "peer", sender)
```
We've lost this log message somehow; can we restore?

patrickhuie19 replied:
358e7b810c1fa356f88b35b019fefbff17c51cc0

krehermann commented on core/capabilities/remote/executable/request/client_request.go:
```diff
@@ -267,3 +291,33 @@ func (c *ClientRequest) sendResponse(response clientResponse) {
+	}
+
+	metadata = resp.Metadata
+	resp.Metadata = commoncap.ResponseMetadata{}
```
why?

patrickhuie19 replied:
If we see something unexpected it unifies the hash calculation to keep the identical response calculation consistent

krehermann commented on core/capabilities/remote/executable/request/client_request_test.go:
```diff
@@ -464,6 +464,81 @@ func Test_ClientRequest_MessageValidation(t *testing.T) {
+		require.NoError(t, err)
+		defer request.Cancel(errors.New("test end"))
+
+		<-dispatcher.msgs
```
whats happening here? code comment pls

patrickhuie19 replied:
inherited from the setup of the `execute` test. removed, it doesn't change anything

krehermann commented on core/capabilities/remote/executable/request/client_request.go:
```diff
@@ -229,25 +228,50 @@ func (c *ClientRequest) OnMessage(_ context.Context, msg *types.MessageBody) err
+
+			nodeReports = append(nodeReports, rpt)
+		} else {
+			c.lggr.Warnf("node metering detail contained %d records; expected exactly 1", len(metadata.Metering))
```
should this be an errorw?

and is this path tested?

patrickhuie19 replied:
yes, thanks, updated to errorw. 

> and is this path tested? 
it's tested in one direction - if no metering data is added, `len(metadata.Metering) = 0`

patrickhuie19 committed:
gauge --> counter; adding labels to logger

patrickhuie19 committed:
removing counter increment when unecessary
